pkg/types/binary_test.go

```diff
@@ -0,0 +1,25 @@
+package types
+
+import (
+	"encoding/json"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestBinary_MarshalJSON(t *testing.T) {
+	assertBinaryMarshalJSON(t, nil, nil)
+	assertBinaryMarshalJSON(t, make(Binary, 0, 1), nil)
+	assertBinaryMarshalJSON(t, Binary(" "), "20")
+}
+
+func assertBinaryMarshalJSON(t *testing.T, input Binary, expect any) {
+	t.Helper()
+
+	jsn, err := input.MarshalJSON()
+	require.NoError(t, err)
+
+	var actual any
+	require.NoError(t, json.Unmarshal(jsn, &actual))
+
+	require.Equal(t, expect, actual)
```

Comment:
I also find it strange that you unmarshal here instead of just providing the expected JSON as a string.

Reply:
The JSON's whitespace can vary, but that's irrelevant. Go values are more expressive and syntax-highlighted.

Reply:
There are no JSON arrays or objects involved here, so where should there be spaces?

Reply:
Also, I want the JSON to be valid/decodable at all in the first place, that code verifies it for me.

Reply:
I mean part of what makes this look strange to me is that it uses `Unmarshal()`, but into type `any` instead of `Binary`. And so far, there is no test for `Unmarshal()`, so my suggestion would actually be a test for both: `Binary` as input, `Marshal()`, check that it matches an expected string, `Unmarshal()` into a new `Binary`, check that it's equal to the first one. Tests everything you want and as a bonus, also adds tests for `Binary.UnmarshalJSON()`.

Reply:
> There are no JSON arrays or objects involved here, so where should there be spaces?

You know what, you're right. My mistake.